crates/rpc/src/v05/method/get_transaction_status.rs

```diff
@@ -0,0 +1,298 @@
+            finality_status,
+            execution_status: Some(receipt.execution_status.into()),
+        }))
+    })
+    .await
+    .context("Joining database task")??;
+
+    if let Some(db_status) = db_status {
+        return Ok(db_status);
+    }
+
+    // Check gateway for rejected transactions.
+    use starknet_gateway_client::GatewayApi;
+    context
+        .sequencer
+        .transaction(input.transaction_hash)
+        .await
+        .context("Fetching transaction from gateway")
+        .and_then(|tx| {
+            use starknet_gateway_types::reply::transaction_status::FinalityStatus as GatewayFinalityStatus;
+            use starknet_gateway_types::reply::transaction_status::ExecutionStatus as GatewayExecutionStatus;
+
+            match (tx.finality_status, tx.execution_status) {
+                (_, GatewayExecutionStatus::Rejected) => Ok(GetTransactionStatusOutput {
+                    finality_status: FinalityStatus::Rejected,
+                    execution_status: None,
+                }),
+                (GatewayFinalityStatus::Received, _) => Ok(GetTransactionStatusOutput {
+                    finality_status: FinalityStatus::Received,
+                    execution_status: None,
+                }),
+                (finality_status, GatewayExecutionStatus::Reverted) => Ok(GetTransactionStatusOutput {
+                    finality_status: finality_status.try_into()?,
+                    execution_status: Some(ExecutionStatus::Reverted),
+                }),
+                (finality_status, GatewayExecutionStatus::Succeeded) => Ok(GetTransactionStatusOutput {
+                    finality_status: finality_status.try_into()?,
+                    execution_status: Some(ExecutionStatus::Succeeded),
+                }),
+            }
+        })
+        .map_err(|_| GetTransactionStatusError::TxnHashNotFoundV04)
+}
+
+async fn pending_status(
+    pending: &PendingData,
+    tx_hash: &TransactionHash,
+) -> Option<GetTransactionStatusOutput> {
+    use starknet_gateway_types::reply::transaction::ExecutionStatus as GatewayExecutionStatus;
+
+    pending
+        .block()
+        .await
+        .map(|block| {
+            block.transaction_receipts.iter().find_map(|rx| {
+                if &rx.transaction_hash == tx_hash {
+                    if rx.execution_status == GatewayExecutionStatus::Reverted {
+                        Some(GetTransactionStatusOutput {
+                            finality_status: FinalityStatus::AcceptedOnL2,
+                            execution_status: Some(ExecutionStatus::Reverted),
+                        })
+                    } else {
+                        Some(GetTransactionStatusOutput {
+                            finality_status: FinalityStatus::AcceptedOnL2,
+                            execution_status: Some(ExecutionStatus::Succeeded),
+                        })
+                    }
+                } else {
+                    None
+                }
```

Comment:
Maybe some variation of? I also prefer matching on the execution status to ensure its an exhaustive check (in case other variants are added in the future). Could also consider `impl From<GatewayExecutionStatus> for ExecutionStatus`.
```suggestion
                (rx.transaction_hash == tx_hash)
                    .then(|rx| {
                        let execution_status = match rx.execution_status {
                            GatewayExecutionStatus::Reverted => ExecutionStatus::Reverted,
                            GatewayExecutionStatus::Succeeded=> ExecutionStatus::Succeeded,
                        };
                        GetTransactionStatusOutput {
                            finality_status: FinalityStatus::AcceptedOnL2,
                            execution_status,
                        }
                    }
```

Reply:
Implementing `From<GatewayExecutionStatus>` isn't possible because `GatewayExecutionStatus` also has `Rejected` that we can't map to the RPC execution status.

And it's a similar story with `match` too. While the type itself does have `Rejected` it's impossible for a transaction that shows up in the pending block to be `Rejected` so having that in the match would just complicate things?

Reply:
Hmm. Right. Maybe we should be more careful there then? The only place we can get a `Rejected` status is from the gateway's transaction status method - so maybe we should instead only add it there?